src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,414 @@
+    Err(CredentialError::retryable_from_str("No token or error in the cache. This should never happen. Something has gone wrong. Open an issue at https://github.com/googleapis/google-cloud-rust/issues/new?template=bug_report.md."))
+}
+
+impl<T: TokenProvider> TokenCache<T> {
+    // TODO(#1210) - implementation in progress
+    #[allow(dead_code)]
+    pub fn new(inner: T) -> TokenCache<T> {
+        TokenCache {
+            token: Arc::new(RwLock::new(initial_token())),
+            refresh_in_progress: Arc::new(Mutex::new(())),
+            refresh_notify: Arc::new(Notify::new()),
+            inner: Arc::new(inner),
+        }
+    }
+
+    fn state(&self) -> TokenState {
+        let token = self.token.read().unwrap();
+        match &*token {
+            Ok(t) => match t.expires_at {
+                None => TokenState::Valid(t.clone()),
+                Some(e) => {
+                    let now = Instant::now().into_std();
+                    if e <= now {
+                        TokenState::Invalid(self.try_refresh_lock())
+                    } else {
+                        TokenState::Valid(t.clone())
+                    }
+                }
+            },
+            Err(_) => TokenState::Invalid(self.try_refresh_lock()),
+        }
+    }
+
+    fn try_refresh_lock(&self) -> RefreshLock {
+        self.refresh_in_progress.clone().try_lock_owned()
+    }
+}
+
+type RefreshLock = std::result::Result<tokio::sync::OwnedMutexGuard<()>, tokio::sync::TryLockError>;
+
+enum TokenState {
+    Invalid(RefreshLock),
+    Valid(Token),
+}
+
+#[async_trait::async_trait]
+impl<T: TokenProvider + 'static> TokenProvider for TokenCache<T> {
+    async fn get_token(&self) -> Result<Token> {
+        match self.state() {
+            TokenState::Valid(token) => {
+                // We have a valid token, and it is not expiring any time soon.
+                // Return it immediately.
+                Ok(token)
+            }
+
+            TokenState::Invalid(refresh_guard) => {
+                match refresh_guard {
+                    // Check if there are any outstanding refreshes...
+                    Ok(guard) => {
+                        // No refreshes. We should start one.
+                        let token = self.inner.get_token().await;
+
+                        // Store the token, or an updated error.
+                        *self.token.write().unwrap() = token.clone();
+
+                        // The refresh is complete. Release the refresh guard.
+                        drop(guard);
+
+                        // Notify any and all waiters.
+                        self.refresh_notify.notify_waiters();
+
+                        // Return here without asking for the token lock again.
+                        token
+                    }
+                    Err(_) => {
+                        // There is already a refresh. We will await its result.
+                        self.refresh_notify.notified().await;
```

Comment:
Hope it's okay for me to comment, I'm concerned here; is there any guarantee that this happens before `notify_waiters` above? I think in theory, the guarded code could run, call `notify_waiters`, and then `notified` here has no waker.